group/assignment6/RearrangeCars.cpp

```diff
@@ -0,0 +1,56 @@
+#include"RearrangeCars.h"
+
+RearrangeCars::RearrangeCars(const vector<int> &startStates) : startStates(startStates){}
+
+void RearrangeCars::setStartStates(const vector<int> &startStates){
+	this->startStates = startStates;
+}
+
+int RearrangeCars::findEmptySlotId(){
+	for (int slotId = 0; slotId < this->startStates.size(); slotId++){
+		if (this->startStates[slotId] == -1) return slotId; // no car has -1 id, it is reserved for empty slot
+	}
+	return -1; // This code should not be reached
+}
+
+vector<int> RearrangeCars::findCarPositions(const vector<int> &currentStates){
+	vector<int> carPositions(currentStates.size() - 1);
+	for (int i = 0; i < currentStates.size(); i++)
+	{
+		if (currentStates[i] != -1) // check for empty slot
+			carPositions[currentStates[i]] = i;
+	}
+	return carPositions;
+}
+
+int RearrangeCars::findMisplacedCar(const vector<int> &carPosition, const vector<int> &endStates){
+	for (int carId = 0; carId < carPosition.size(); carId++){
+		if (endStates[carPosition[carId]] != carId) return carId;
+	}
+	return -1; // If every car is on its slot
+}
+
+void RearrangeCars::moveCar(vector<int> &currentStates, vector<int> &carPosition, int from, int to){
+	currentStates[to] = currentStates[from];
+	currentStates[from] = -1;
+	carPosition[currentStates[to]] = to;
+}
+
+void RearrangeCars::generateSequenceOfMoves(const vector<int> &endStates, vector<Move> &sequenceOfMoves){

```

Comment:
Maybe bool should be returned instead of void (false of the invalid input case)?

Reply:
I agree, so method which validates input vectors will be called at the beginning of this function, not in a test before calling this function?
I think that calling that method here will be a better option.

Reply:
I think so. We shouldn't let user think about validation, better to do it here for certain.